Enable backend gutter marks for lambda functions

sdubov opened the pull request:
## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

## Description
- Add AWS Rider backend C# logic
- Update gradle to compile AWS Rider backend logic and copy it into plugin output
- Add Rider Protocol extensions to handle backend events
- Update gradle to generate protocol models on Kotlin and C# sides
- Add fake backend icons for actions in Lambda gutter mark context
- Add Icons patcher on frontend to update fake backend icons with frontend icons
- Pass credentials when opening run lambda run configuration
- Renable the Configuration type instance to separate from configuration type
- Minor updates

## Motivation and Context
Gutter marks allow users to run/debug lambda from editor from a gutter context menu. To set a gutter mark to a correct lambda method we should parse PSI to determine a right function that represents a lambda. In Rider all information about PSI lives on a backend. To make a gutter mark work in Rider, we should enable a backend in plugin and show a backend gutter mark with backend actions.

Since we have all actions defined on a frontend we should fire event on user interactions with context icons that are defined on a backend. This is enabled by Rider protocol extension that is listening for a backend gutter mark popup menu event and fire action on a frontend.

## Related Issue(s)
None

## Testing
```
JetBrains Rider 2019.2 EAP
Runtime version: 1.8.0_212-release-1566-b8 x86_64
VM: OpenJDK 64-Bit Server VM by JetBrains s.r.o
macOS 10.14.5
GC: PS Scavenge, PS MarkSweep
Memory: 480M
Cores: 8
```
- Run/debug/create actions from a gutter mark in Rider
- Verify new Gradle tasks
- Verify Idea plugin gutter marks

## Screenshots (if appropriate)
<img width="652" alt="LambdaGutterMark" src="https://user-images.githubusercontent.com/6064345/61903224-ad27cf80-af2c-11e9-92b8-257d171defc2.png">

## Checklist
- [x] I have read the **README** document
- [x] I have read the **CONTRIBUTING** document
- [x] Local run of `gradlew check` succeeds
- [x] My code follows the code style of this project
- [ ] I have added tests to cover my changes
- [x] All new and existing tests passed
- [ ] A short description of the change has been added to the **CHANGELOG**

## License
I confirm that my contribution is made under the terms of the Apache 2.0 license.

sdubov committed:
Enable backend gutter marks for lambda functions

- Add AWS Rider backend C# logic
- Update gradle to compile AWS Rider backend logic and copy it into plugin output
- Add Rider Protocol extensions to handle backend events
- Update gradle to generate protocol models on Kotlin and C# sides
- Add fake backend icons for actions in Lambda gutter mark context
- Add Icons patcher on frontend to update fake backend icons with frontend icons
- Pass credentials when opening run lambda run configuration
- Renable the Configuration type instance to separate from configuration type
- Minor updates

kashmervil commented on jetbrains-rider/ReSharper.AWS/Directory.Build.props:
```diff
@@ -0,0 +1,23 @@
+  </PropertyGroup>
+  
+  <PropertyGroup>
+    <FantomasVersion>[2.9.9]</FantomasVersion>
```
This is not needed unless you're aiming to use some F#

kashmervil commented on jetbrains-rider/ReSharper.AWS/src/AWS.Daemon/RunMarkers/LambdaRunMarkerGutterMark.cs:
```diff
@@ -0,0 +1,85 @@
+      var solution = Shell.Instance.GetComponent<SolutionsManager>().Solution;
+      if (solution == null) yield break;
+
+      switch (runMarker.AttributeId)
```
I'd better check the attribute with an `if (runMarker.AttributeId != LambdaRunMarkerAttributeIds.LAMBDA_RUN_METHOD_MARKER_ID) yield break;`

Unless you have plans to add more `Id`s

kashmervil commented on jetbrains-rider/ReSharper.AWS/src/AWS.Daemon/RunMarkers/LambdaRunMarkerProvider.cs:
```diff
@@ -0,0 +1,68 @@
+        {
+            if (!(file is ICSharpFile csharpFile)) return;
+
+//            var canBeTaskReturning = EntryPointUtil.CanBeTaskReturning(csharpFile);
```
Leftover

kashmervil commented on jetbrains-rider/ReSharper.AWS/src/AWS.Daemon/RunMarkers/LambdaRunMarkerProvider.cs:
```diff
@@ -0,0 +1,68 @@
+                   method.Parameters.Count <= 2 &&
+                   method.Parameters.Any(parameter =>
+//                       Equals(parameter.Type, GetType("Amazon.Lambda.Core.ILambdaContext")));
+                       (parameter.Type as IDeclaredType)?.GetClrName().FullName == "Amazon.Lambda.Core.ILambdaContext");
```
It's better to create a ClrName for "Amazon.Lambda.Core.ILambdaContext", put it in a static field and then compare it with the `ClrName` of the type

kashmervil commented on jetbrains-rider/ReSharper.AWS/src/AWS.Daemon/RunMarkers/LambdaRunMarkerProvider.cs:
```diff
@@ -0,0 +1,68 @@
+            return method.Parameters.Count > 0 &&
+                   method.Parameters.Count <= 2 &&
+                   method.Parameters.Any(parameter =>
+//                       Equals(parameter.Type, GetType("Amazon.Lambda.Core.ILambdaContext")));
```
Leftover

kashmervil commented on jetbrains-rider/ReSharper.AWS/src/AWS.Daemon/RunMarkers/LambdaRunMarkerGutterMark.cs:
```diff
@@ -0,0 +1,85 @@
+
+    private string ComposeHandlerString(RunMarkerHighlighting runMarker)
+    {
+      var projectDirectory = runMarker.Project.ProjectFileLocation.Parent.NameWithoutExtension;
```
It's better to check the corner cases

kashmervil commented on jetbrains-rider/ReSharper.AWS/src/AWS.Daemon/RunMarkers/LambdaRunMarkerGutterMark.cs:
```diff
@@ -0,0 +1,85 @@
+    {
+      var projectDirectory = runMarker.Project.ProjectFileLocation.Parent.NameWithoutExtension;
+      var methodName = runMarker.Method.ShortName;
+      var typeString = runMarker.FullName.Substring(0, runMarker.FullName.Length - methodName.Length - 1);
```
Nullability

kashmervil commented on jetbrains-rider/ReSharper.AWS/src/AWS.Daemon/RunMarkers/LambdaRunMarkerGutterMark.cs:
```diff
@@ -0,0 +1,85 @@
+      var lambdaHost = solution.GetComponent<LambdaHost>();
+
+      var methodName = runMarker.Method.ShortName;
+      var handlerString = ComposeHandlerString(runMarker);
```
No nullability checks for `methodName` which I believe can be the case e.g. for 
`class Foo 
{
  void (ILambdaContext context)
  {
  }
}
`
R# parser tries to recover from errors and to provide the tree anyway, so some properties could be null.

kashmervil commented on jetbrains-rider/src/software/aws/toolkits/jetbrains/services/lambda/dotnet/DotNetLambdaHandlerResolver.kt:
```diff
@@ -10,10 +10,11 @@ import com.intellij.psi.NavigatablePsiElement
 import com.intellij.psi.PsiElement
 import com.intellij.psi.search.GlobalSearchScope
 import com.jetbrains.rd.framework.impl.RpcTimeouts
+import com.jetbrains.rd.framework.impl.startAndAdviseSuccess
```
Is `DotNetLambdaHandlerResolver` really needed after moving to a backend line marker providing?

kashmervil commented on jetbrains-rider/src/software/aws/toolkits/jetbrains/services/lambda/dotnet/DotNetLambdaHandlerResolver.kt:
```diff
@@ -100,6 +86,39 @@ class DotNetLambdaHandlerResolver : LambdaHandlerResolver {
+     */
+    override fun shouldShowLineMarker(handler: String): Boolean = false
+
+    fun getFieldIdByHandlerName(project: Project, handler: String): Int {
```
Let's pass all the needed information from the backend as a struct containng `className , methodName, targetFramework, projectId` instead of `string`?

kashmervil replied:
Is this API needed in the first place?

kashmervil commented on jetbrains-rider/src/software/aws/toolkits/jetbrains/services/lambda/upload/DotNetLambdaLineMarker.kt:
```diff
@@ -0,0 +1,21 @@
+// Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package software.aws.toolkits.jetbrains.services.lambda.upload
```
And this? Why should we keep it?

kashmervil commented on jetbrains-rider/src/software/aws/toolkits/jetbrains/utils/RuntimeUtil.kt:
```diff
@@ -0,0 +1,46 @@
+import software.aws.toolkits.jetbrains.services.lambda.validOrNull
+import java.io.IOException
+
+object RuntimeUtil {
```
Isn't there an API for it already?

kashmervil replied:
Looks really ad-hoc

sdubov committed:
Code review fixes and updated logic to define lambda functions

- Disable DotNetLineMarker on a frontend
- Update logic to define lambda method
- Remove redundant comments
- Update logic with lambda function detection
- Add frontend tests to verify gutter marks
